### PR TITLE
[QE/PoC] Cypress integration

### PR DESCRIPTION
**Pull request by ScriptingShrimp**
This pull request draft (PR) affects associates beyond our immediate team, so this is why I build on top of the Open Decision Framework. Determined by the nature of this proposal, this place serves as the only place where progress will be published. The goal here is to make this PR as transparent and inclusive as possible, also this space is right for a free exchange of critical ideas. Any collaboration is highly encouraged and welcomed - together, we can do more. 

**TL;DR:**

> Not a real PR, just discussion. Selenium bad, Cypress good - Introduction to a new tool, PoC.

**The problem statement:**
Efforts to maintain the current UI regression suite are demanding. This is caused by:
- Steep learning curve for maintenance and extending the UI suite
- Nearly none involvement form UI developers team, sporadic involvement of QE department - common denominator is knowledge transfer 
- Stability of tests caused by their high complexity

**Possible approaches:**
1. Refactor current UI suite: decrease complexity - fewer interactions and assertions per test case (atomic testing) and invest time in knowledge transfer: how to maintain, debug and fix tests between developers and QEs.
1. Introduce a simple tool with a flat learning curve, so it's more likely to be adopted by programmers and also by non programmers. Focus on stability, speed and atomic testing.

**Suggested solution:**

- Restate product requirements for E2E and UI testing. 
- Split the regression UI testing into more comprehensible layers. Partially refactor the current suite.
- Identify and classify (test) cases ideal for Cypress PoC - create tight integration within product repository. 

**What this is:**
- Example of how Cypress should be integrated - Proof of Concept
- Discussion about what Cypress tool is, what it can cover
- Place for knowledge transfer, place for highlighting best practices using Cypress
- Forum to raise questions, concerns and positive or negative feedback on UI regression
- Playground for routine exercise around git ecosystem

**What this is not:**
- A real PR 
- Discussion about withdrawing selenium support

### CODE REVIEW

**Comment by tsze-redhat** [comment]
I spoke with Pavel and turns out, this is intended to be discussion forum for automation for both Maistra and Kiali.
So, everyone, please comment.

Also, included in Pavel's comments above is the topic of E2E vs UI testing.
I would like to encourage everyone to step outside of their assigned area and learn about how team members test their area.

Please chime in and post your comments here and keep this discussion going. We should keep the momentum going and figure out the best fit of the tools we have (Selenium / Cypress / CIs and various scripts).

Thank you!

File: cypress/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+`node -v
+v14.18.1`
+
+`npm -v
+6.14.15`
+
+last install yarn package manager
+
+`sudo npm install --global yarn`
```
Comment (jshaughn) [inline cypress/README.md:31]: nodejs, npm and yarn are already required for kiali-ui, so maybe we should just defer to the versions used for the ui itself.

File: cypress/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+clone kiali-ui repository and cd into directory
+
+`git clone ...`
+
+`cd kiali-ui`
+
+checkout on right branch
+
+`git checkout cypress-integration`
```
Comment (jshaughn) [inline cypress/README.md:41]: Is this branch going to always exist?  Probably we should just be using the PR branch?

**Review by jshaughn — COMMENTED** [review]
I don't have any issue with QE introducing cypress as a testing tool.  A couple of comments that can likely be ignored given that this is not a real PR.  Actualoly, this should likely be a Draft PR.  My only questions is about whether there is any additional instumentation needed in the ui code to support Cypress testing, and whether the tests themselves would be a QE effort, or would involve the dev team?   I expect devs could certainly contribute new tests, I was unsure whether it was proposed to be mandatory.

File: cypress/README.md (hunk 1 above)
Comment (ScriptingShrimp) [reply cypress/README.md:31]: No doubts there, you are completely right.

**Comment by ScriptingShrimp** [comment]
> I don't have any issue with QE introducing cypress as a testing tool. A couple of comments that can likely be ignored given that this is not a real PR. Actualoly, this should likely be a Draft PR. 

converted back to draft PR. I believe you can't add reviewers to draft PR.  

> My only questions is about whether there is any additional instumentation needed in the ui code to support Cypress testing,

no additional instumentation is needed, in general - cypress is self-contained  

> and whether the tests themselves would be a QE effort, or would involve the dev team? I expect devs could certainly contribute new tests, I was unsure whether it was proposed to be mandatory.

the rule of the thumb here is: with help of the developers, test coverage scales better

**Comment by jshaughn** [comment]
> the rule of the thumb here is: with help of the developers, test coverage scales better

To be clear, I would expect e2e and automation to remain in the domain of QE.  Dev team would, as we have in the past, help to facilitate testing with any reasonable code-support, but likely would not devote a lot of time to that sort of test development.   If Cypress was to be seen as useful for dev-side unit testing, I could imagine it perhaps being used in that way by the dev team, to complement existing unit testing.   Also, I would expect the existing Selenium tests to be maintained, right, even if Cypress is introduced.

**Comment by ScriptingShrimp** [comment]
> To be clear, I would expect e2e and automation to remain in the domain of QE. Dev team would, as we have in the past, help to facilitate testing with any reasonable code-support, but likely would not devote a lot of time to that sort of test development. 

 - E2E and automation in general is domain of the QE's. 
 - Design of the testcase and its ownership (i.e. maintenance) is domain of the QE's.
 - Technical implementation of the test case itself and code ownership related to that - we need to discuss more in detail. The honest answer here: It is something we can not afford to own right now. Not yet.

After approving PoC, we will start to take small, incremental steps. Increasing the coverage, run E2E tests locally. Make the tests and the automation overall rock solid. Reiterate. With that proceed to CI. 

This whole process is something really critical for us to understand, and any assistance is welcomed.
@prachiyadav @pbajjuri20 @skondkar ^^ please, I want to hear your thoughts on that. If you don't understand something, or you disagree with statements above, it is the right place to let us all know. You ale also part of the decision making

**Comment by lucasponce** [comment]
@ScriptingShrimp, thanks for the PR.

This will help to give visibility of the discussion between QE and Dev.

I think that it requires first an iteration between the QE team as it's more a discussion between Selenium vs Cypress where QE have experience, from Dev we can't add too much on that sense.

I have some considerations as I see some similarities between the current jest snapshots that we use, recording snapshots is not easy to maintain, and reviewing the PoC the main difference may be the "easy" way to record the backend, but tests are similar as Jest snapshots provide and we are reconsidering them for unit testing.

If those may help in the E2E is a different discussion, but my first early feedback, but as commented, I'm just talking from a Dev perspective.

**Comment by pbajjuri20** [comment]
```

Here, the difference between **Cypress VS Selenium** 

Selenium VS Cypress:

Features                                                Selenium                                                            Cypress
 
1. Architecture                                   1. Driver/Server                                                       1. No Server

2. Supported Languages                  2. Java, JavaScript, C#, Python, Ruby			  2. JavaScript

3. Selectors					 3. Name, ID, Class Name, Lint Text, Tag                 3. JQuery
								 Name, CSS Selector Xpath												


4. Frameworks Support                   4. Support multiple Frameworks based on specific         4. Mocha JS, Cucumber
							programming launages (eg: Junit for Java Cucumber
							 for Java Script														

5. Speed    					5. slow										5. Fast

6. Screeshots					6.  supported						                        6. supported

7. Video Recording				7. supported							          	7. supported

8. Community Support			8. well estabilished documentation 		         	8. rapidly growing community
 

```

**Comment by pbajjuri20** [comment]
Compared to Selenium, we have more advantages with Cypress. I'm comfortable with Cypress and I Agree with @ScriptingShrimp. I have done some POC with Mocha JS script on Installing the Operators and navigating the Dev Console. It is much faster than Selenium.

**Comment by lucasponce** [comment]
Hi,
Just ping on this PR to see what's the status of the discussion, next steps.

**Comment by ScriptingShrimp** [comment]
Closing this draft due to inactivity. PoC was successfully presented across QE's, informal feedback were overall positive. 

This draft served as introduction and discussion about Cypress. If you have more general questions, feel free to contact me, or reopen this draft PR. 

Upcoming actions: 
- smaller PR's with 
Later there will be "real" PR with more specific scope. 

Meanwhile its possible that tool get merged as a part of more technical PR, but with very limited scope.

**Comment by ScriptingShrimp** [comment]
Closing this draft due to inactivity. PoC was successfully presented across QE's, informal feedback were overall positive.

Upcoming actions:
- smaller PR's with more technical details
- creating plan for seamless integration into QE's workflows

Meanwhile its possible that tool get merged as a part of more technical PR, but with very limited scope.

This draft served as introduction and discussion about Cypress. If you have more general questions, feel free to contact me, or reopen this draft PR.

**Pull request closed without merging**